### DISCUSSION TITLE
[WASM] Multithread and HttpClient

### DISCUSSION THREAD

**Discussion by sarosigy** [unoplatform/uno#9891] [Q&A]
Downloaded the latest .NET 7 (rc) and trying out the multithreading support.
In the csproj file:
```
...
    <TargetFramework>net7.0</TargetFramework>
...
  <PropertyGroup>
    <WasmShellEnableThreads>true</WasmShellEnableThreads>
  </PropertyGroup>
  <ItemGroup>
    <WasmShellExtraEmccFlags Include="-s PTHREAD_POOL_SIZE=8"/>
  </ItemGroup>
...
```

The app compiles well, and runs as expected, until the following code reached:
```
            System.Net.Http.HttpClient cli = new System.Net.Http.HttpClient();
            try
            {
                var res = await cli.PostAsync("http://localhost:7878/api/Cargo/count",
                    new System.Net.Http.StringContent("getthis", System.Text.Encoding.UTF8, "application/json"));
                if (res.IsSuccessStatusCode)
                {
                    ;
                }
                else
                {
                    ;
                }
            }
            catch (Exception ex)
            {
                ;
            }
```
And an Exception is caught: 
ERR13: Invalid JS object handle '4' while invoking 'abort'

The number may vary.

**Should it work at all, or not?**

I read some vague comment that strict CORS is needed to use WASM threading, but in this case it is obviously not met, because the api created and hosted is not on the same site with the self-hosted Uno WASM project.

Of course, with no multithreading (commented out csproj items) the app runs with no exceptions.
And the rc of the .NET 7 is better than the preview, because with the preview the await postasync just hung and never returned.

**Comment by jeromelaban** [comment]
Thanks for the report. Can you tell which version of the bootstrapper you are using ?

**Reply by jeromelaban** [reply]
(please continue here)

yes, there's a version 7.0-dev as of yesterday, which targets .NET 7.0 RC1. It contains many threading-related fixes.

**Reply by sarosigy** [reply]
I'm trying right now...

**Reply by sarosigy** [reply]
I have lost the ability to stop on a breakpoint. All breakpoints set are gray in vscode.
The app still works with no threads.
With threads, the error message is different now:

dotnet.js:5 Error: Assert failed: ERR41: Unknown converter for type 13
mono_wasm_bind_js_function @ dotnet.js:5
_mono_wasm_bind_js_function @ dotnet.js:14
$func344 @ dotnet.wasm:0x23804
$func312 @ dotnet.wasm:0x22a65
$func248 @ dotnet.wasm:0x13d06
$func247 @ dotnet.wasm:0x12ba4
$func8257 @ dotnet.wasm:0x1b2cf9
$func2118 @ dotnet.wasm:0x92371
$func2116 @ dotnet.wasm:0x922dc
$func2283 @ dotnet.wasm:0x98ccb
invokeEntryPoint @ dotnet.js:14
self.onmessage @ dotnet.worker.js:1

**Reply by sarosigy** [reply]
Text of the exception caught and displayed:

![Screenshot_20220919_203336](https://user-images.githubusercontent.com/37872626/191090361-0df2c4c6-6f80-410b-bc3e-013fa7cf26a7.jpg)

**Reply by jeromelaban** [reply]
Thanks. This is a repro you should be able to open directly on the .NET runtime repo. If you can tag me there, it'll be appreciated.

**Reply by sarosigy** [reply]
OK, I create a small repro project to open an issue there.

**Reply by sarosigy** [reply]
I created the repro and opened an issue: https://github.com/dotnet/runtime/issues/76063

**Comment by sarosigy** [comment]
I use these package versions:
```
    <PackageReference Include="Uno.WinUI.WebAssembly" Version="4.6.0-dev.168" />
    <PackageReference Include="Uno.Wasm.Bootstrap" Version="4.0.0-dev.193" />
    <PackageReference Include="Uno.Wasm.Bootstrap.DevServer" Version="4.0.0-dev.193" />

```
Is there a package with major version 7?
